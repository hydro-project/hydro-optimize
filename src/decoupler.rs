use std::cell::RefCell;
use std::collections::HashMap;
use std::rc::Rc;

use hydro_lang::builder::ir::{
    DebugInstantiate, DebugType, HydroIrMetadata, HydroIrOpMetadata, HydroNode, HydroRoot, TeeNode,
    transform_bottom_up, traverse_dfir,
};
use hydro_lang::live_collections::stream::networking::{
    deserialize_bincode_with_type, serialize_bincode_with_type,
};
use hydro_lang::location::MemberId;
use hydro_lang::location::dynamic::LocationId;
use proc_macro2::Span;
use serde::{Deserialize, Serialize};
use stageleft::quote_type;
use syn::visit_mut::VisitMut;

use crate::repair::{cycle_source_to_sink_input, inject_id, inject_location};
use crate::rewrites::ClusterSelfIdReplace;

#[derive(Clone, Serialize, Deserialize)]
pub struct Decoupler {
    pub output_to_decoupled_machine_after: Vec<usize>, /* The output of the operator at this index should be sent to the decoupled machine */
    pub output_to_original_machine_after: Vec<usize>, /* The output of the operator at this index should be sent to the original machine */
    pub place_on_decoupled_machine: Vec<usize>, /* This operator should be placed on the decoupled machine. Only for sources */
    pub orig_location: LocationId,
    pub decoupled_location: LocationId,
}

fn add_network(node: &mut HydroNode, new_location: &LocationId) {
    let metadata = node.metadata().clone();
    let output_debug_type = metadata.output_type.clone().unwrap();

    let parent_id = metadata.location_kind.root().raw_id();
    let node_content = std::mem::replace(node, HydroNode::Placeholder);

    // Map from b to (MemberId, b), where MemberId is the id of the decoupled (or original) node we're sending to
    let ident = syn::Ident::new(
        &format!("__hydro_lang_cluster_self_id_{}", parent_id),
        Span::call_site(),
    );
    let f: syn::Expr = syn::parse_quote!(|b| (
        ::hydro_lang::location::MemberId::<()>::from_raw(#ident),
        b
    ));
    let cluster_id_type = quote_type::<MemberId<()>>();
    let mapped_output_type: syn::Type = syn::parse_quote!((#cluster_id_type, #output_debug_type));
    let mapped_node = HydroNode::Map {
        f: f.into(),
        input: Box::new(node_content),
        metadata: HydroIrMetadata {
            location_kind: metadata.location_kind.root().clone(), // Remove any ticks
            output_type: Some(DebugType(Box::new(mapped_output_type.clone()))),
            cardinality: None,
            tag: None,
            op: HydroIrOpMetadata {
                backtrace: metadata.op.backtrace.clone(),
                cpu_usage: None,
                network_recv_cpu_usage: None,
                id: None,
            },
        },
    };

    // Set up the network node
    let output_type = output_debug_type.clone().0;
    let network_node = HydroNode::Network {
        serialize_fn: Some(serialize_bincode_with_type(true, &output_type)).map(|e| e.into()),
        instantiate_fn: DebugInstantiate::Building,
        deserialize_fn: Some(deserialize_bincode_with_type(
            Some(&quote_type::<()>()),
            &output_type,
        ))
        .map(|e| e.into()),
        input: Box::new(mapped_node),
        metadata: HydroIrMetadata {
            location_kind: new_location.clone(),
            output_type: Some(DebugType(Box::new(mapped_output_type))),
            cardinality: None,
            tag: None,
            op: HydroIrOpMetadata {
                backtrace: metadata.op.backtrace.clone(),
                cpu_usage: None,
                network_recv_cpu_usage: None,
                id: None,
            },
        },
    };

    // Map again to remove the member Id (mimicking send_anonymous)
    let f: syn::Expr = syn::parse_quote!(|(_, b)| b);
    let mapped_node = HydroNode::Map {
        f: f.into(),
        input: Box::new(network_node),
        metadata: HydroIrMetadata {
            location_kind: new_location.clone(),
            output_type: Some(output_debug_type),
            cardinality: None,
            tag: None,
            op: HydroIrOpMetadata {
                backtrace: metadata.op.backtrace.clone(),
                cpu_usage: None,
                network_recv_cpu_usage: None,
                id: None,
            },
        },
    };
    *node = mapped_node;
}

fn add_tee(
    node: &mut HydroNode,
    new_location: &LocationId,
    new_inners: &mut HashMap<(usize, LocationId), Rc<RefCell<HydroNode>>>,
) {
    let metadata = node.metadata().clone();
    let inner_id = if let HydroNode::Tee { inner, .. } = node {
        inner.0.borrow().op_metadata().id.unwrap()
    } else {
        std::panic!("Decoupler add_tee() called on non-Tee");
    };

    let new_inner = new_inners
        .entry((inner_id, new_location.clone()))
        .or_insert_with(|| {
            println!(
                "Adding network before Tee to location {:?} after id: {}",
                new_location, inner_id
            );
            add_network(node, new_location);
            let node_content = std::mem::replace(node, HydroNode::Placeholder);
            Rc::new(RefCell::new(node_content))
        })
        .clone();

    let teed_node = HydroNode::Tee {
        inner: TeeNode(new_inner),
        metadata,
    };
    *node = teed_node;
}

fn decouple_node(
    node: &mut HydroNode,
    decoupler: &Decoupler,
    next_stmt_id: &mut usize,
    new_inners: &mut HashMap<(usize, LocationId), Rc<RefCell<HydroNode>>>,
) {
    // Replace location of sources, if necessary
    if decoupler.place_on_decoupled_machine.contains(next_stmt_id) {
        match node {
            HydroNode::Source { metadata, .. } | HydroNode::Network { metadata, .. } => {
                println!(
                    "Changing source/network destination from {:?} to location {:?}, id: {}",
                    metadata.location_kind,
                    decoupler.decoupled_location.clone(),
                    next_stmt_id
                );
                metadata
                    .location_kind
                    .swap_root(decoupler.decoupled_location.clone());
            }
            _ => {
                std::panic!(
                    "Decoupler placing non-source/network node on decoupled machine: {}",
                    node.print_root()
                );
            }
        }
        return;
    }

    // Otherwise, replace where the outputs go
    let new_location = if decoupler
        .output_to_decoupled_machine_after
        .contains(next_stmt_id)
    {
        &decoupler.decoupled_location
    } else if decoupler
        .output_to_original_machine_after
        .contains(next_stmt_id)
    {
        &decoupler.orig_location
    } else {
        return;
    };

    match node {
        HydroNode::Placeholder | HydroNode::Network { .. } => {
            std::panic!(
                "Decoupler modifying placeholder node or incorrectly handling network node: {}",
                next_stmt_id
            );
        }
        HydroNode::Tee { .. } => {
            println!(
                "Creating a TEE to location {:?}, id: {}",
                new_location, next_stmt_id
            );
            add_tee(node, new_location, new_inners);
        }
        _ => {
            println!(
                "Creating network to location {:?} after node {}, id: {}",
                new_location,
                node.print_root(),
                next_stmt_id
            );
            add_network(node, new_location);
        }
    }
}

fn fix_cluster_self_id_root(root: &mut HydroRoot, mut locations: ClusterSelfIdReplace) {
    if let ClusterSelfIdReplace::Decouple {
        decoupled_cluster_id,
        ..
    } = locations
        && root.input_metadata()[0].location_kind.root().raw_id() == decoupled_cluster_id
    {
        root.visit_debug_expr(|expr| {
            locations.visit_expr_mut(&mut expr.0);
        });
    }
}

fn fix_cluster_self_id_node(node: &mut HydroNode, mut locations: ClusterSelfIdReplace) {
    if let ClusterSelfIdReplace::Decouple {
        decoupled_cluster_id,
        ..
    } = locations
        && node.metadata().location_kind.root().raw_id() == decoupled_cluster_id
    {
        node.visit_debug_expr(|expr| {
            locations.visit_expr_mut(&mut expr.0);
        });
    }
}

pub fn decouple(ir: &mut [HydroRoot], decoupler: &Decoupler) {
    let mut new_inners = HashMap::new();
    traverse_dfir(
        ir,
        |_, _| {},
        |node, next_stmt_id| {
            decouple_node(node, decoupler, next_stmt_id, &mut new_inners);
        },
    );

    // Fix IDs since we injected nodes
    inject_id(ir);
    // Fix locations since we changed some
    let cycle_source_to_sink_input = cycle_source_to_sink_input(ir);
    inject_location(ir, &cycle_source_to_sink_input);
    // Fix CLUSTER_SELF_ID for the decoupled node
    let locations = ClusterSelfIdReplace::Decouple {
        orig_cluster_id: decoupler.orig_location.raw_id(),
        decoupled_cluster_id: decoupler.decoupled_location.raw_id(),
    };
    transform_bottom_up(
        ir,
        &mut |leaf| {
            fix_cluster_self_id_root(leaf, locations);
        },
        &mut |node| {
            fix_cluster_self_id_node(node, locations);
        },
        true,
    );
}

#[cfg(test)]
mod tests {
    use std::collections::HashSet;

    use hydro_deploy::Deployment;
    use hydro_lang::builder::rewrites::persist_pullup::persist_pullup;
    use hydro_lang::builder::{FlowBuilder, ir};
    use hydro_lang::location::Location;
    use hydro_lang::nondet::nondet;
    use stageleft::q;

    use crate::debug::name_to_id_map;
    use crate::decoupler::{Decoupler, decouple};
    use crate::repair::inject_id;

    fn decouple_mini_program<'a>(
        output_to_decoupled_machine_after: Vec<(&str, i32)>, // name, offset
        output_to_original_machine_after: Vec<(&str, i32)>,
        place_on_decoupled_machine: Vec<(&str, i32)>,
    ) -> (
        hydro_lang::location::Cluster<'a, ()>,
        hydro_lang::location::Cluster<'a, ()>,
        hydro_lang::location::Cluster<'a, ()>,
        hydro_lang::builder::built::BuiltFlow<'a>,
    ) {
        let builder = FlowBuilder::new();
        let send_cluster = builder.cluster::<()>();
        let recv_cluster = builder.cluster::<()>();
        let decoupled_cluster = builder.cluster::<()>();

        send_cluster
            .source_iter(q!(0..10))
            .map(q!(|a| a + 1))
            .ir_node_named("map")
            .broadcast_bincode(&recv_cluster, nondet!(/** test */))
            .values()
            .for_each(q!(|a| println!("Got it: {}", a)));

        let built = builder
            .optimize_with(persist_pullup)
            .optimize_with(inject_id)
            .optimize_with(|ir| {
                // Convert named nodes to IDs, accounting for the offset
                let name_to_id = name_to_id_map(ir);
                let decoupler = Decoupler {
                    output_to_decoupled_machine_after: output_to_decoupled_machine_after
                        .into_iter()
                        .map(|(name, offset)| {
                            (name_to_id.get(name).cloned().unwrap() as i32 + offset) as usize
                        })
                        .collect(),
                    output_to_original_machine_after: output_to_original_machine_after
                        .into_iter()
                        .map(|(name, offset)| {
                            (name_to_id.get(name).cloned().unwrap() as i32 + offset) as usize
                        })
                        .collect(),
                    place_on_decoupled_machine: place_on_decoupled_machine
                        .into_iter()
                        .map(|(name, offset)| {
                            (name_to_id.get(name).cloned().unwrap() as i32 + offset) as usize
                        })
                        .collect(),
                    decoupled_location: decoupled_cluster.id().clone(),
                    orig_location: send_cluster.id().clone(),
                };
                decouple(ir, &decoupler)
            });
        (send_cluster, recv_cluster, decoupled_cluster, built)
    }

    async fn check_decouple_mini_program(
        output_to_decoupled_machine_after: Vec<(&str, i32)>, // name, offset
        output_to_original_machine_after: Vec<(&str, i32)>,
        place_on_decoupled_machine: Vec<(&str, i32)>,
    ) {
        let (send_cluster, recv_cluster, decoupled_cluster, built) = decouple_mini_program(
            output_to_decoupled_machine_after,
            output_to_original_machine_after,
            place_on_decoupled_machine,
        );

        // Check outputs
        let mut deployment = Deployment::new();
        let nodes = built
            .with_cluster(&send_cluster, vec![deployment.Localhost(); 1])
            .with_cluster(&recv_cluster, vec![deployment.Localhost(); 3])
            .with_cluster(&decoupled_cluster, vec![deployment.Localhost(); 1])
            .deploy(&mut deployment);

        deployment.deploy().await.unwrap();

        let recv_members = nodes.get_cluster(&recv_cluster).members();
        let mut stdouts = vec![];
        for member in recv_members {
            use hydro_lang::deploy::DeployCrateWrapper;

            stdouts.push(member.stdout().await);
        }

        deployment.start().await.unwrap();

        for mut stdout in stdouts {
            let mut expected = HashSet::new();
            let mut received = HashSet::new();
            for i in 1..11 {
                received.insert(stdout.recv().await.unwrap());
                expected.insert(format!("Got it: {}", i));
            }
            assert_eq!(expected, received);
        }
    }

    #[test]
    fn decouple_after_source_ir() {
        let output_to_decoupled_machine_after = vec![("map", -1)];
        let output_to_original_machine_after = vec![];
        let place_on_decoupled_machine = vec![
            // the source of cluster membership (TODO(shadaj): should have a better way of identifying)
            ("map", -6),
        ];

        let built = decouple_mini_program(
            output_to_decoupled_machine_after,
            output_to_original_machine_after,
            place_on_decoupled_machine,
        )
        .3;

        ir::dbg_dedup_tee(|| {
            hydro_build_utils::assert_debug_snapshot!(built.ir());
        });
    }

    #[tokio::test]
    async fn decouple_after_source() {
        let output_to_decoupled_machine_after = vec![("map", -1)];
        let output_to_original_machine_after = vec![];
        let place_on_decoupled_machine = vec![
            // the source of cluster membership (TODO(shadaj): should have a better way of identifying)
            ("map", -6),
        ];

        check_decouple_mini_program(
            output_to_decoupled_machine_after,
            output_to_original_machine_after,
            place_on_decoupled_machine,
        )
        .await
    }

    #[test]
    fn move_source_decouple_map_ir() {
        let output_to_decoupled_machine_after = vec![];
        let output_to_original_machine_after = vec![("map", 0)];
        let place_on_decoupled_machine = vec![("map", -1)];

        let built = decouple_mini_program(
            output_to_decoupled_machine_after,
            output_to_original_machine_after,
            place_on_decoupled_machine,
        )
        .3;

        ir::dbg_dedup_tee(|| {
            hydro_build_utils::assert_debug_snapshot!(built.ir());
        });
    }

    #[tokio::test]
    async fn move_source_decouple_map() {
        let output_to_decoupled_machine_after = vec![];
        let output_to_original_machine_after = vec![("map", 0)];
        let place_on_decoupled_machine = vec![("map", -1)];

        check_decouple_mini_program(
            output_to_decoupled_machine_after,
            output_to_original_machine_after,
            place_on_decoupled_machine,
        )
        .await
    }
}
