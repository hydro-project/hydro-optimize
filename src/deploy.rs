use std::collections::HashMap;
use std::sync::Arc;

use hydro_deploy::gcp::GcpNetwork;
use hydro_deploy::rust_crate::tracing_options::{DEBIAN_PERF_SETUP_COMMAND, TracingOptions};
use hydro_deploy::{Deployment, Host};
use hydro_lang::deploy::TrybuildHost;
use tokio::sync::RwLock;

pub struct ReusableHosts {
    pub hosts: HashMap<String, Arc<dyn Host>>, // Key = display_name
    pub host_arg: String,
    pub project: String,
    pub network: Arc<RwLock<GcpNetwork>>,
}

impl ReusableHosts {
    // NOTE: Creating hosts with the same display_name in the same deployment will result in undefined behavior.
    fn lazy_create_host(
        &mut self,
        deployment: &mut Deployment,
        display_name: String,
    ) -> Arc<dyn Host> {
        self.hosts
            .entry(display_name.clone())
            .or_insert_with(|| {
                if self.host_arg == "gcp" {
                    deployment
                        .GcpComputeEngineHost()
                        .project(&self.project)
                        .machine_type("n2-standard-4")
                        .image("debian-cloud/debian-12")
                        .region("us-central1-c")
                        .network(self.network.clone())
                        .display_name(display_name)
                        .add()
                } else {
                    deployment.Localhost()
                }
            })
            .clone()
    }

    pub fn get_process_hosts(
        &mut self,
        deployment: &mut Deployment,
        display_name: String,
    ) -> TrybuildHost {
        let rustflags = if self.host_arg == "gcp" {
            "-C opt-level=3 -C codegen-units=1 -C strip=none -C debuginfo=2 -C lto=off -C link-args=--no-rosegment"
        } else {
            "-C opt-level=3 -C codegen-units=1 -C strip=none -C debuginfo=2 -C lto=off"
        };
        TrybuildHost::new(self.lazy_create_host(deployment, display_name.clone()))
            .additional_hydro_features(vec!["runtime_measure".to_string()])
            .build_env(
                "HYDRO_RUNTIME_MEASURE_CPU_PREFIX",
                super::deploy_and_analyze::CPU_USAGE_PREFIX,
            )
            .rustflags(rustflags)
            .tracing(
                TracingOptions::builder()
                    .perf_raw_outfile(format!("{}.perf.data", display_name.clone()))
                    .fold_outfile(format!("{}.data.folded", display_name))
                    .frequency(128)
                    .setup_command(DEBIAN_PERF_SETUP_COMMAND)
                    .build(),
            )
    }

    pub fn get_cluster_hosts(
        &mut self,
        deployment: &mut Deployment,
        cluster_name: String,
        num_hosts: usize,
    ) -> Vec<TrybuildHost> {
        (0..num_hosts)
            .map(|i| self.get_process_hosts(deployment, format!("{}{}", cluster_name, i)))
            .collect()
    }
}
