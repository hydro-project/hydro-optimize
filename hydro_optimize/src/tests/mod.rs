use std::cell::RefCell;
use std::collections::HashMap;

use hydro_build_utils::insta;
use hydro_lang::compile::rewrites::persist_pullup;
use hydro_lang::deploy::HydroDeploy;
use hydro_lang::location::Location;
use hydro_lang::prelude::*;

use crate::decoupler;
use crate::partitioner::Partitioner;

mod two_pc;

struct DecoupledCluster;

#[test]
fn decoupled_compute_pi_ir() {
    let builder = FlowBuilder::new();
    let (cluster, _) = hydro_test::cluster::compute_pi::compute_pi(&builder, 8192);
    let decoupled_cluster = builder.cluster::<DecoupledCluster>();
    let decoupler = decoupler::Decoupler {
        output_to_decoupled_machine_after: vec![4],
        output_to_original_machine_after: vec![],
        place_on_decoupled_machine: vec![],
        decoupled_location: decoupled_cluster.id().clone(),
        orig_location: cluster.id().clone(),
    };
    let multi_run_metadata = RefCell::new(vec![]);
    let built = builder
        .optimize_with(persist_pullup::persist_pullup)
        .optimize_with(|roots| decoupler::decouple(roots, &decoupler, &multi_run_metadata, 0))
        .into_deploy::<HydroDeploy>();

    insta::assert_debug_snapshot!(built.ir());

    for (id, ir) in built.preview_compile().all_dfir() {
        insta::with_settings!({
            snapshot_suffix => format!("surface_graph_{id}"),
        }, {
            insta::assert_snapshot!(ir.surface_syntax_string());
        });
    }
}

#[test]
fn partitioned_simple_cluster_ir() {
    let builder = FlowBuilder::new();
    let (_, cluster) = hydro_test::cluster::simple_cluster::simple_cluster(&builder);
    let partitioner = Partitioner {
        nodes_to_partition: HashMap::from([(5, vec!["1".to_string()])]),
        num_partitions: 3,
        location_id: cluster.id().raw_id(),
        new_cluster_id: None,
    };
    let built = builder
        .optimize_with(persist_pullup::persist_pullup)
        .optimize_with(|roots| crate::partitioner::partition(roots, &partitioner))
        .into_deploy::<HydroDeploy>();

    insta::assert_debug_snapshot!(built.ir());

    for (id, ir) in built.preview_compile().all_dfir() {
        insta::with_settings!({
            snapshot_suffix => format!("surface_graph_{id}")
        }, {
            insta::assert_snapshot!(ir.surface_syntax_string());
        });
    }
}
