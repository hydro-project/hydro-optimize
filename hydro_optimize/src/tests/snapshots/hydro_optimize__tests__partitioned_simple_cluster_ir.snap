---
source: hydro_optimize/src/tests/mod.rs
expression: built.ir()
---
[
    ForEach {
        f: stageleft :: runtime_support :: fn1_type_hint :: < (hydro_test :: __staged :: __deps :: hydro_lang :: location :: member_id :: MemberId < () > , (hydro_test :: __staged :: __deps :: hydro_lang :: location :: member_id :: MemberId < () > , i32)) , () > ({ use hydro_test :: __staged :: __deps :: * ; use hydro_test :: __staged :: cluster :: simple_cluster :: * ; | (id , d) | println ! ("node received: ({}, {:?})" , id , d) }),
        input: Map {
            f: | (sender_id , b) | (:: hydro_lang :: location :: MemberId :: < _ > :: from_raw (sender_id . raw_id / 3usize as u32) , b),
            input: Network {
                serialize_fn: Some(
                    :: hydro_lang :: runtime_support :: stageleft :: runtime_support :: fn1_type_hint :: < (hydro_test :: __staged :: __deps :: hydro_lang :: location :: member_id :: MemberId < () > , i32) , _ > (| data | { hydro_lang :: runtime_support :: bincode :: serialize (& data) . unwrap () . into () }),
                ),
                instantiate_fn: <network instantiate>,
                deserialize_fn: Some(
                    | res | { let (id , b) = res . unwrap () ; (hydro_lang :: location :: MemberId :: < () > :: from_raw (id) , hydro_lang :: runtime_support :: bincode :: deserialize :: < (hydro_test :: __staged :: __deps :: hydro_lang :: location :: member_id :: MemberId < () > , i32) > (& b) . unwrap ()) },
                ),
                input: Inspect {
                    f: stageleft :: runtime_support :: fn1_borrow_type_hint :: < (hydro_test :: __staged :: __deps :: hydro_lang :: location :: member_id :: MemberId < () > , i32) , () > ({ use hydro_test :: __staged :: __deps :: * ; use hydro_test :: __staged :: cluster :: simple_cluster :: * ; let CLUSTER_SELF_ID__free = hydro_lang :: location :: MemberId :: < () > :: from_raw ({ __hydro_lang_cluster_self_id_1 / 3usize as u32 }) ; move | n | println ! ("cluster received: {:?} (self cluster id: {})" , n , CLUSTER_SELF_ID__free) }),
                    input: Network {
                        serialize_fn: Some(
                            :: hydro_lang :: runtime_support :: stageleft :: runtime_support :: fn1_type_hint :: < (hydro_lang :: location :: MemberId < _ > , (hydro_test :: __staged :: __deps :: hydro_lang :: location :: member_id :: MemberId < () > , i32)) , _ > (| (id , data) | { (id . raw_id , hydro_lang :: runtime_support :: bincode :: serialize (& data) . unwrap () . into ()) }),
                        ),
                        instantiate_fn: <network instantiate>,
                        deserialize_fn: Some(
                            | res | { hydro_lang :: runtime_support :: bincode :: deserialize :: < (hydro_test :: __staged :: __deps :: hydro_lang :: location :: member_id :: MemberId < () > , i32) > (& res . unwrap ()) . unwrap () },
                        ),
                        input: Map {
                            f: stageleft :: runtime_support :: fn1_type_hint :: < (hydro_test :: __staged :: __deps :: hydro_lang :: location :: member_id :: MemberId < () > , i32) , (hydro_test :: __staged :: __deps :: hydro_lang :: location :: member_id :: MemberId < () > , (hydro_test :: __staged :: __deps :: hydro_lang :: location :: member_id :: MemberId < () > , i32)) > ({ use hydro_test :: __staged :: __deps :: * ; use hydro_test :: __staged :: cluster :: simple_cluster :: * ; | (id , n) | (id , (id , n)) }),
                            input: Map {
                                f: | (orig_dest , struct_or_tuple) : (:: hydro_lang :: location :: MemberId < _ > , _) | { let mut s = :: std :: hash :: DefaultHasher :: new () ; :: std :: hash :: Hash :: hash (& struct_or_tuple . 1 , & mut s) ; let partition_val = :: std :: hash :: Hasher :: finish (& s) as u32 ; (:: hydro_lang :: location :: MemberId :: < () > :: from_raw ((orig_dest . raw_id * 3usize as u32) + (partition_val % 3usize as u32) as u32) , struct_or_tuple) },
                                input: Delta {
                                    inner: CrossProduct {
                                        left: Persist {
                                            inner: FilterMap {
                                                f: stageleft :: runtime_support :: fn1_type_hint :: < (hydro_test :: __staged :: __deps :: hydro_lang :: location :: member_id :: MemberId < () > , hydro_test :: __staged :: __deps :: hydro_lang :: location :: MembershipEvent) , core :: option :: Option < hydro_test :: __staged :: __deps :: hydro_lang :: location :: member_id :: MemberId < () > > > ({ use hydro_test :: __staged :: __deps :: * ; use hydro_test :: __staged :: cluster :: simple_cluster :: * ; | (i , e) | match e { MembershipEvent :: Joined => Some (i) , MembershipEvent :: Left => None , } }),
                                                input: Map {
                                                    f: stageleft :: runtime_support :: fn1_type_hint :: < & hydro_test :: __staged :: __deps :: hydro_lang :: location :: member_id :: MemberId < () > , (hydro_test :: __staged :: __deps :: hydro_lang :: location :: member_id :: MemberId < () > , hydro_test :: __staged :: __deps :: hydro_lang :: location :: MembershipEvent) > ({ use hydro_lang :: __staged :: __deps :: * ; use hydro_lang :: __staged :: location :: * ; | id | (* id , MembershipEvent :: Joined) }),
                                                    input: Source {
                                                        source: Iter(
                                                            { use hydro_lang :: __staged :: __deps :: * ; use hydro_lang :: __staged :: location :: * ; let underlying_memberids__free = { let all_ids = unsafe { :: std :: mem :: transmute :: < _ , & [hydro_lang :: location :: MemberId < () >] > (__hydro_lang_cluster_ids_1) } ; & all_ids [0 .. all_ids . len () / 3usize] } ; underlying_memberids__free },
                                                        ),
                                                        metadata: HydroIrMetadata {
                                                            location_kind: Process(
                                                                0,
                                                            ),
                                                            output_type: Some(
                                                                & hydro_test :: __staged :: __deps :: hydro_lang :: location :: member_id :: MemberId < () >,
                                                            ),
                                                        },
                                                    },
                                                    metadata: HydroIrMetadata {
                                                        location_kind: Process(
                                                            0,
                                                        ),
                                                        output_type: Some(
                                                            (hydro_test :: __staged :: __deps :: hydro_lang :: location :: member_id :: MemberId < () > , hydro_test :: __staged :: __deps :: hydro_lang :: location :: MembershipEvent),
                                                        ),
                                                    },
                                                },
                                                metadata: HydroIrMetadata {
                                                    location_kind: Process(
                                                        0,
                                                    ),
                                                    output_type: Some(
                                                        hydro_test :: __staged :: __deps :: hydro_lang :: location :: member_id :: MemberId < () >,
                                                    ),
                                                },
                                            },
                                            metadata: HydroIrMetadata {
                                                location_kind: Process(
                                                    0,
                                                ),
                                                output_type: Some(
                                                    hydro_test :: __staged :: __deps :: hydro_lang :: location :: member_id :: MemberId < () >,
                                                ),
                                            },
                                        },
                                        right: Persist {
                                            inner: Source {
                                                source: Iter(
                                                    { use hydro_test :: __staged :: __deps :: * ; use hydro_test :: __staged :: cluster :: simple_cluster :: * ; 0 .. 5 },
                                                ),
                                                metadata: HydroIrMetadata {
                                                    location_kind: Process(
                                                        0,
                                                    ),
                                                    output_type: Some(
                                                        i32,
                                                    ),
                                                },
                                            },
                                            metadata: HydroIrMetadata {
                                                location_kind: Process(
                                                    0,
                                                ),
                                                output_type: Some(
                                                    i32,
                                                ),
                                            },
                                        },
                                        metadata: HydroIrMetadata {
                                            location_kind: Process(
                                                0,
                                            ),
                                            output_type: Some(
                                                (hydro_test :: __staged :: __deps :: hydro_lang :: location :: member_id :: MemberId < () > , i32),
                                            ),
                                        },
                                    },
                                    metadata: HydroIrMetadata {
                                        location_kind: Process(
                                            0,
                                        ),
                                        output_type: Some(
                                            (hydro_test :: __staged :: __deps :: hydro_lang :: location :: member_id :: MemberId < () > , i32),
                                        ),
                                    },
                                },
                                metadata: HydroIrMetadata {
                                    location_kind: Process(
                                        0,
                                    ),
                                    output_type: Some(
                                        (hydro_test :: __staged :: __deps :: hydro_lang :: location :: member_id :: MemberId < () > , i32),
                                    ),
                                },
                            },
                            metadata: HydroIrMetadata {
                                location_kind: Process(
                                    0,
                                ),
                                output_type: Some(
                                    (hydro_test :: __staged :: __deps :: hydro_lang :: location :: member_id :: MemberId < () > , (hydro_test :: __staged :: __deps :: hydro_lang :: location :: member_id :: MemberId < () > , i32)),
                                ),
                            },
                        },
                        metadata: HydroIrMetadata {
                            location_kind: Cluster(
                                1,
                            ),
                            output_type: Some(
                                (hydro_test :: __staged :: __deps :: hydro_lang :: location :: member_id :: MemberId < () > , i32),
                            ),
                        },
                    },
                    metadata: HydroIrMetadata {
                        location_kind: Cluster(
                            1,
                        ),
                        output_type: Some(
                            (hydro_test :: __staged :: __deps :: hydro_lang :: location :: member_id :: MemberId < () > , i32),
                        ),
                    },
                },
                metadata: HydroIrMetadata {
                    location_kind: Process(
                        0,
                    ),
                    output_type: Some(
                        (hydro_test :: __staged :: __deps :: hydro_lang :: location :: member_id :: MemberId < () > , (hydro_test :: __staged :: __deps :: hydro_lang :: location :: member_id :: MemberId < () > , i32)),
                    ),
                },
            },
            metadata: HydroIrMetadata {
                location_kind: Process(
                    0,
                ),
                output_type: Some(
                    (hydro_test :: __staged :: __deps :: hydro_lang :: location :: member_id :: MemberId < () > , (hydro_test :: __staged :: __deps :: hydro_lang :: location :: member_id :: MemberId < () > , i32)),
                ),
            },
        },
        op_metadata: HydroIrOpMetadata,
    },
]
